design/pen.behavior.md

```diff
@@ -0,0 +1,60 @@
+# Pen Behavior
+
+The Pen behavior is a method of recording the path by which a user moves a grabbed object so that various scenearios 
+can be carried out that wish to capture the virtual drawing of a user.  This behavior will carry with it client side visuals
+to ensure low-latency feedback, but will supply the complete curve data needed by the MRE app to acto on once the drawing is done.
+
+## Architecture
+
+Create a new behavior that will enable grabbable on the actor that the behavior is added to.  This behavior will then allow the user
+to execute a primary action to begin recording the transform changes over time to use it to build a spline curve.  The MRE runtime will
+generate a wet ink effect and a mesh to apply to this spline curve.  Callbacks will occur for when the actor with the behavior has been
+picked up and dropped, as well as when the drawing has started and stopped.  There will be transform data sent up to the MRE at a set 
+frequency while the drawing is active to facilitate spline generation by the MRE app.
+
+### Behavior
+``` ts
+/**
+ * Button behavior class containing the target behavior actions.
+ */
+export class PenBehavior extends TargetBehavior {
+	private _hold: DiscreteAction = new DiscreteAction();
+	private _drawing: DiscreteAction = new DiscreteAction();
+
+	/** @inheritdoc */
+	public get behaviorType(): BehaviorType { return 'pen'; }
+
+	public get hold() { return this._hold; }
+	public get drawing() { return this._drawing; }
+
+	public onDrawDataReceived((data: Transform[]) => void);
```

Comment:
I don't see a material difference between the pen and interactable behaviors. `onDrawDataReceived` would be very useful to have on `InteractableBehavior` (but renamed to something less domain specific). If it were added there, what differences remain?

Reply:
I agree - it seems like PenBehavior is InteractableBehavior+wet ink effect+drawDataReceived 
If wet ink effect and drawDataReceived were optional features it could be a single behavior.


---
In reply to: [376066644](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/478#discussion_r376066644) [](ancestors = 376066644)

Reply:
Soren had liked the idea of a separate more generic behavior.  What I am thinking about more now is having a `ToolBehavior` which the `PenBehavior` extends.  I definitely can see other behaviors wanting to extend something like tool and add additional functionality to the `ToolBehavior`.  I will update the design to reflect that with the `onHeldPathReceived` API in the base class `ToolBehavior`.